File: 10-1.rb
#!/usr/bin/env ruby

open = [ '(', '[', '{', '<' ]
close = [ ')', ']', '}', '>' ]
match = {
  '(' => ')',
  '[' => ']',
  '{' => '}',
  '<' => '>'
}

score = {
  ')' => 3,
  ']' => 57,
  '}' => 1197,
  '>' => 25137
}

lines = File.read('10.input').lines.map(&:strip).map { |l| l.split '' }

errors = []

lines.each do |line|
  stack = []
  line.each do |token|
    if open.any? token
      stack.push token
    elsif close.any? token
      top = stack[-1]
      if token == match[top]
        stack.pop
      else
        errors.push token
        break
      end
    end
  end
end

print errors.map { |token| score[token] }.sum
print "\n"

File: 10-1.py
#!/usr/bin/env python

from collections import deque

open_tokens = [ '(', '[', '{', '<' ]
close_tokens = [ ')', ']', '}', '>' ]
match = {
  '(': ')',
  '[': ']',
  '{': '}',
  '<': '>'
}

score = {
  ')': 3,
  ']': 57,
  '}': 1197,
  '>': 25137
}

lines = [line.strip() for line in open('10.input').readlines()]
lines = [list(line) for line in lines]

errors = []

for line in lines:
    stack = deque([])
    for token in line:
        if token in open_tokens:
            stack.append(token)
        elif token in close_tokens:
            top = stack[-1]
            if token == match[top]:
                stack.pop()
            else:
                errors.append(token)
                break

print(sum([score[token] for token in errors]))
